Added clustered etcd example

mfojtik opened the pull request:
(still heavily in progress)

mfojtik commented:
@smarterclayton @danmcp @bparees still missing README and some polishing but I would appreciate early feedback :-)

mfojtik commented:
updated README

rhcarvalho commented on examples/etcd/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+RUN ETCD_URL=https://github.com/coreos/etcd/releases/download/${ETCD_RELEASE}/etcd-${ETCD_RELEASE}-linux-amd64.tar.gz && \
+  mkdir -p /tmp/etcd && cd /tmp/etcd && \
+  curl -L ${ETCD_URL} -o etcd.tar.gz && \
+  tar -xzf  etcd.tar.gz --strip-components=1 && \
```
Why not stream from `curl` to `tar`?

    curl -L ${ETCD_URL} | tar -xzf --strip-components=1

mfojtik replied:
yeah I should do that... less number of steps then. fixed.

rhcarvalho commented on examples/etcd/README.md:
```diff
@@ -0,0 +1,71 @@
+
+https://github.com/coreos/etcd
+
+### Pre-requires
```
Prerequirements?

rhcarvalho commented on examples/etcd/README.md:
```diff
@@ -0,0 +1,71 @@
+
+### How to use this template
+
+You can import this template to Openshift using:
```
s/Openshift/OpenShift

rhcarvalho commented on examples/etcd/README.md:
```diff
@@ -0,0 +1,71 @@
+### How does it work
+
+This template creates two Services. The first service is used for initial
+discovery and stores informations about running members. This service is used
```
s/informations/information ;)

rhcarvalho commented on examples/etcd/README.md:
```diff
@@ -0,0 +1,71 @@
+members you will have to start this pod again manually.
+
+The 'etcd' replication controller manage creation of the etcd cluster members.
+by default this template will start 3 members. The members then register using
```
/s/by/By

soltysh replied:
s/register using/register themselves using/

soltysh commented on examples/etcd/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+FROM centos:centos7
+MAINTAINER http://openshift.io
```
Either not set that field, or set it to your name.

mfojtik replied:
https://github.com/eliskasl/container-best-practices/commit/3be9e3fecaf93b6c5f60d11f0d6bedea0cd5c58b#commitcomment-10921606

soltysh commented on examples/etcd/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+
+RUN ETCD_URL=https://github.com/coreos/etcd/releases/download/${ETCD_RELEASE}/etcd-${ETCD_RELEASE}-linux-amd64.tar.gz && \
+  mkdir -p /tmp/etcd && cd /tmp/etcd && \
+  curl -L ${ETCD_URL} | tar -xzf - --strip-components=1 && \
```
You're using this value only once, what's the reason for setting it?

mfojtik replied:
yo make the line shorter and more readable, I don't think it is a big deal to leave it there....

rhcarvalho replied:
+1 for setting ETCD_URL, makes it easier to read / maintain.

soltysh commented on examples/etcd/README.md:
```diff
@@ -0,0 +1,71 @@
+
+The 'etcd' replication controller manage creation of the etcd cluster members.
+by default this template will start 3 members. The members then register using
+the discovery service and elect the leader. You adjust the number of replicas as
```
s/You adjust/You can adjust/

soltysh commented on examples/etcd/README.md:
```diff
@@ -0,0 +1,71 @@
+The 'etcd' replication controller manage creation of the etcd cluster members.
+by default this template will start 3 members. The members then register using
+the discovery service and elect the leader. You adjust the number of replicas as
+long as the 'etcd-discovery' service is running.
```
Maybe it'll be worth showing the command/steps to do that?

mfojtik replied:
i think we have now couple ways to do that so I will leave that for documentation ;-)

soltysh commented on examples/etcd/README.md:
```diff
@@ -0,0 +1,71 @@
+
+### Cleaning up
+
+If you're done playing, you can remove all created resources by executing
```
s/remove all created/remove all the created/

soltysh commented:
A couple of small nits, but generally LGTM.

mfojtik commented:
@soltysh @rhcarvalho thanks guys!

derekwaynecarr commented:
I am interested in updating the example of how to run OpenShift master as pod on Kubernetes:
https://github.com/GoogleCloudPlatform/kubernetes/tree/master/examples/openshift-origin

I don't see how I could support any of the security flags... don't we need a --ca-file, --cert-file, --key-file?  All of this seems like it would be necessary to support running an openshift master as a pod on GKE.

eparis commented:
@aveshagarwal

aveshagarwal commented on examples/etcd/etcd.sh:
```diff
@@ -0,0 +1,46 @@
+  -advertise-client-urls http://${address}:2379 \
+  -listen-client-urls http://${address}:2379 \
+  -data-dir /var/lib/etcd \
+  -name ${member_id}
```
@mfojtik  it seems that it missing following:  -discovery ${ETCD_DISCOVERY}

aveshagarwal commented on examples/etcd/template.json:
```diff
@@ -0,0 +1,196 @@
+                },
+                {
+                  "name": "ETCD_INITIAL_CLUSTER_TOKEN",
+                  "value": "${ETCD_CLUSTER_TOKEN}"
```
I think for etcd discovery container, ETCD_INITIAL_CLUSTER_STATE and ETCD_INITIAL_CLUSTER_TOKEN might be omitted.

mfojtik replied:
@aveshagarwal yes, I have to test that, but I think you are correct.

aveshagarwal commented on examples/etcd/etcd.sh:
```diff
@@ -0,0 +1,46 @@
+
+echo "Discovered members: ${endpoints}"
+echo "Adjusting etcd cluster size to ${peers_count}"
+curl -sX PUT ${ETCD_DISCOVERY}/_config/size -d value=${peers_count}
```
@mfojtik  It seems that you are updating cluster size here and adding new etcd node. Am I right? If so are you sure it is possible to do, i.e. have you tested it? As far as my understanding goes, you could set cluster size once for boot strapping and after the cluster is booted with that size, you need to add the new etcd node manually.

mfojtik replied:
@aveshagarwal I tested it and when you raise the number, then the new node is picked up and it force to do new election of the 'leader'. It does not work 100% time as you have to have quorum. Also sometimes container starts very fast the the DNS lookup does not return all members... We can add retry or some sleep to give it enough time.

aveshagarwal replied:
@mfojtik I am wondering if you need to really use both DNS and local discovery service. I think you could just use one of them, and infact could avoid race conditions with your current design. 
If you just use local discovery service, you could get the number of nodes from the template by specifying it as one of the parameters (equal to number of replicas). Or you could just use DNS for bootstrapping etcd cluster by using DNS SRV records.

mfojtik replied:
@aveshagarwal yeah, I like the DNS SRV approach more than having environment value defined in the PodTemplate->Container with the current number of replicas. That way you will have to update the RC definition everytime you want to resize your cluster, right?

mfojtik replied:
@aveshagarwal just talked with Clayton on IRC, we want to have static number of replicas set for etcd (like 3.) and we will not allow to scale up. In case there is a failure, the RC will start a new Pod to replace failed Pod and the cluster will elect a new leader.

mfojtik commented:
@aveshagarwal I updated the example to use 3 replicas by default and stick to this number. I reworked the template to also use DeploymentConfig(s). I also tested failure case (docker kill one of the etcd-member), in that case a replacement Pod is launched automatically and join the cluster.

@smarterclayton @bparees @danmcp I think we can merge this as it is now if you guys don't have any objections.

bparees commented on examples/etcd/teardown.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+osc delete all -l template=etcd
```
if this is all this does, why not get rid of this script and just document the command in the readme?

bparees commented:
one suggested script cleanup and lgtm.

mfojtik commented:
@bparees thanks! updated.

mfojtik commented:
[merge]
